File: packages/tracecast-ts/src/integrations/llm.ts
import { Tracer, getCurrentTrace } from "../core/tracer";
import { Span, SpanType } from "../types";
import { calculateCost } from "../core/costCalculator";
import { randomUUID } from "crypto";

let _defaultTracer: Tracer | null = null;

export function setDefaultTracer(tracer: Tracer): void {
  _defaultTracer = tracer;
}

export function getDefaultTracer(): Tracer | null {
  return _defaultTracer;
}

function resolveTracer(): Tracer {
  return _defaultTracer ?? new Tracer();
}

function extractTokens(response: any, provider: string): { input: number; output: number; cached: number } {
  if (provider === "openai") {
    const usage = response?.usage ?? {};
    const details = usage?.prompt_tokens_details ?? {};
    return {
      input: usage?.prompt_tokens ?? 0,
      output: usage?.completion_tokens ?? 0,
      cached: details?.cached_tokens ?? 0,
    };
  }
  if (provider === "anthropic") {
    const usage = response?.usage ?? {};
    return {
      input: usage?.input_tokens ?? 0,
      output: usage?.output_tokens ?? 0,
      cached: usage?.cache_read_input_tokens ?? 0,
    };
  }
  return { input: 0, output: 0, cached: 0 };
}

function extractContent(response: any, provider: string): string | undefined {
  if (provider === "openai") {
    try {
      const choices = response?.choices ?? [];
      if (choices.length) {
        const msg = choices[0]?.message;
        if (msg) return msg.content ?? msg.reasoning_content ?? "";
      }
    } catch { return undefined; }
  }
  if (provider === "anthropic") {
    try {
      const content = response?.content ?? [];
      const texts: string[] = [];
      for (const block of content) {
        if (block?.type === "text") texts.push(block.text ?? "");
        else if (block?.type === "thinking") texts.push(`[thinking] ${block.thinking ?? ""}`);
      }
      return texts.length ? texts.join("") : undefined;
    } catch { return undefined; }
  }
  return undefined;
}

function extractInputText(kwargs: Record<string, any>): string | undefined {
  const messages = kwargs?.messages as any[] | undefined;
  if (messages?.length) {
    const last = messages[messages.length - 1];
    if (last?.content) return String(last.content);
  }
  return undefined;
}

export async function traceLLMCall<T>(
  fn: () => Promise<T>,
  opts: {
    provider: string;
    model: string;
    inputText?: string;
    metadata?: Record<string, unknown>;
  },
): Promise<T> {
  const trace = getCurrentTrace();
  if (!trace) return fn();

  const span: Span = {
    spanId: randomUUID(),
    type: SpanType.LLM,
    name: `llm:${opts.model}`,
    model: opts.model,
    startedAt: new Date(),
    input: opts.inputText,
    metadata: opts.metadata ?? {},
  };

  const tracer = resolveTracer();
  const logger = (tracer as any)._logger ?? null;

  if (logger) {
    logger.llmStart(trace.name, opts.model);
  }

  let response: T;
  try {
    response = await fn();
  } catch (err: any) {
    if (logger) logger.llmError(trace.name, opts.model, err.message);
    span.finishedAt = new Date();
    span.metadata = { ...span.metadata, _error: err.message };
    trace.spans.push(span);
    throw err;
  }

  span.finishedAt = new Date();
  const tokens = extractTokens(response, opts.provider);
  span.tokensIn = tokens.input;
  span.tokensOut = tokens.output;
  span.tokensInCached = tokens.cached || undefined;
  span.costUsd = calculateCost(opts.model, tokens.input, tokens.output, undefined, tokens.cached);
  span.output = extractContent(response, opts.provider);
  trace.spans.push(span);

  if (logger) {
    const latencyMs = span.finishedAt.getTime() - span.startedAt.getTime();
    logger.llmEnd(trace.name, {
      model: opts.model,
      tokensIn: tokens.input,
      tokensOut: tokens.output,
      tokensInCached: tokens.cached || undefined,
      costUsd: span.costUsd,
      latencyMs,
    });
  }

  return response;
}

export function wrapOpenAI(client: any): any {
  return new Proxy(client, {
    get(target, prop, receiver) {
      const val = Reflect.get(target, prop, receiver);
      if (prop === "chat") {
        return new Proxy(val, {
          get(innerTarget, innerProp, innerReceiver) {
            const innerVal = Reflect.get(innerTarget, innerProp, innerReceiver);
            if (innerProp === "completions") {
              return new Proxy(innerVal, {
                get(deepTarget, deepProp, deepReceiver) {
                  const deepVal = Reflect.get(deepTarget, deepProp, deepReceiver);
                  if (deepProp === "create") {
                    return async function (this: any, ...args: any[]) {
                      const kwargs = args[0] ?? {};
                      const model = kwargs.model ?? "unknown";
                      const inputText = extractInputText(kwargs);
                      return traceLLMCall(
                        () => deepVal.apply(this, args),
                        { provider: "openai", model, inputText },
                      );
                    };
                  }
                  return deepVal;
                },
              });
            }
            return innerVal;
          },
        });
      }
      return val;
    },
  });
}

export function wrapAnthropic(client: any): any {
  return new Proxy(client, {
    get(target, prop, receiver) {
      const val = Reflect.get(target, prop, receiver);
      if (prop === "messages") {
        return new Proxy(val, {
          get(innerTarget, innerProp, innerReceiver) {
            const innerVal = Reflect.get(innerTarget, innerProp, innerReceiver);
            if (innerProp === "create") {
              return async function (this: any, ...args: any[]) {
                const kwargs = args[0] ?? {};
                const model = kwargs.model ?? "unknown";
                const inputText = extractInputText(kwargs);
                return traceLLMCall(
                  () => innerVal.apply(this, args),
                  { provider: "anthropic", model, inputText },
                );
              };
            }
            return innerVal;
          },
        });
      }
      return val;
    },
  });
}

File: packages/tracecast-py/tracecast/integrations/llm.py
import inspect
import uuid
import json
from datetime import datetime, timezone
from typing import Any, Callable, Optional, TypeVar
from ..core.tracer import Tracer
from ..core.token_counter import extract_tokens, extract_content, extract_input_text
from ..core.cost_calculator import calculate_cost
from ..core.payload import truncate_payload
from ..models.span import Span, SpanType

T = TypeVar("T")


def trace_llm_call(
    fn: Callable[..., T],
    *,
    provider: str,
    model: str,
    input_text: Optional[str] = None,
    metadata: Optional[dict] = None,
) -> T:
    trace = Tracer.current()
    if trace is None:
        return fn()

    span = Span(
        span_id=str(uuid.uuid4()),
        parent_span_id=getattr(Tracer.current_span(), "span_id", None),
        type=SpanType.LLM,
        name=f"llm:{model}",
        model=model,
        started_at=datetime.now(timezone.utc),
        input=truncate_payload(input_text),
        metadata=metadata or {},
    )

    tracer = _resolve_tracer()
    logger = getattr(tracer, "_tc_logger", None)
    if logger:
        logger.llm_start(trace.name, model=model)

    response: Any = None
    error: Optional[str] = None
    try:
        response = fn()
    except Exception as exc:
        error = str(exc)
        if logger:
            logger.llm_error(trace.name, model=model, error=error)
        span.finished_at = datetime.now(timezone.utc)
        span.mark_error(exc)
        trace.spans.append(span)
        raise

    span.finished_at = datetime.now(timezone.utc)
    tokens = extract_tokens(response, provider)
    span.tokens_in = int(tokens["input"])
    span.tokens_out = int(tokens["output"])
    span.tokens_in_cached = int(tokens.get("cached", 0))
    span.cost_usd = calculate_cost(
        model,
        span.tokens_in,
        span.tokens_out,
        tokens_in_cached=span.tokens_in_cached,
    )
    span.output = truncate_payload(extract_content(response, provider))
    trace.spans.append(span)

    if logger:
        latency_ms = None
        if span.finished_at and span.started_at:
            latency_ms = (span.finished_at - span.started_at).total_seconds() * 1000
        logger.llm_end(
            trace.name,
            model=model,
            tokens_in=span.tokens_in,
            tokens_out=span.tokens_out,
            tokens_in_cached=span.tokens_in_cached,
            cost_usd=span.cost_usd,
            latency_ms=latency_ms,
        )

    return response


async def _async_trace_llm_call(
    coro,
    *,
    provider: str,
    model: str,
    input_text: Optional[str] = None,
):
    """Async counterpart of trace_llm_call — properly awaits the API response."""
    trace = Tracer.current()
    if trace is None:
        return await coro

    span = Span(
        span_id=str(uuid.uuid4()),
        parent_span_id=getattr(Tracer.current_span(), "span_id", None),
        type=SpanType.LLM,
        name=f"llm:{model}",
        model=model,
        started_at=datetime.now(timezone.utc),
        input=truncate_payload(input_text),
        metadata={},
    )
    tracer = _resolve_tracer()
    logger = getattr(tracer, "_tc_logger", None)
    if logger:
        logger.llm_start(trace.name, model=model)

    try:
        response = await coro
    except Exception as exc:
        span.finished_at = datetime.now(timezone.utc)
        span.mark_error(exc)
        trace.spans.append(span)
        if logger:
            logger.llm_error(trace.name, model=model, error=str(exc))
        raise

    span.finished_at = datetime.now(timezone.utc)
    tokens = extract_tokens(response, provider)
    span.tokens_in = int(tokens["input"])
    span.tokens_out = int(tokens["output"])
    span.tokens_in_cached = int(tokens.get("cached", 0))
    span.cost_usd = calculate_cost(
        model,
        span.tokens_in,
        span.tokens_out,
        tokens_in_cached=span.tokens_in_cached,
    )
    span.output = truncate_payload(extract_content(response, provider))
    trace.spans.append(span)

    if logger:
        latency_ms = (span.finished_at - span.started_at).total_seconds() * 1000
        logger.llm_end(
            trace.name,
            model=model,
            tokens_in=span.tokens_in,
            tokens_out=span.tokens_out,
            tokens_in_cached=span.tokens_in_cached,
            cost_usd=span.cost_usd,
            latency_ms=latency_ms,
        )
    return response


def _resolve_tracer() -> Tracer:
    from ..decorators import _default_tracer
    return _default_tracer or Tracer()


def wrap_openai(client: Any) -> Any:
    return _ProxyWrapper(client, provider="openai", intercept_attr="chat.completions.create")


def wrap_anthropic(client: Any) -> Any:
    return _ProxyWrapper(client, provider="anthropic", intercept_attr="messages.create")


class _ProxyWrapper:
    __slots__ = ("_client", "_provider", "_intercept_path")

    def __init__(self, client: Any, *, provider: str, intercept_attr: str):
        self._client = client
        self._provider = provider
        self._intercept_path = intercept_attr.split(".")

    def __getattr__(self, name: str) -> Any:
        target = getattr(self._client, name)
        if name == self._intercept_path[0]:
            return _NestedProxy(
                target,
                provider=self._provider,
                intercept_rest=self._intercept_path[1:],
            )
        return target


class _NestedProxy:
    __slots__ = ("_target", "_provider", "_intercept_rest")

    def __init__(self, target: Any, *, provider: str, intercept_rest: list):
        self._target = target
        self._provider = provider
        self._intercept_rest = intercept_rest

    def __getattr__(self, name: str) -> Any:
        target = getattr(self._target, name)
        if self._intercept_rest and name == self._intercept_rest[0]:
            return _NestedProxy(
                target,
                provider=self._provider,
                intercept_rest=self._intercept_rest[1:],
            )
        return target

    def __call__(self, *args: Any, **kwargs: Any) -> Any:
        model = kwargs.get("model", "unknown")
        input_text = extract_input_text(kwargs, self._provider)
        raw = self._target(*args, **kwargs)
        if inspect.isawaitable(raw):
            return _async_trace_llm_call(
                raw,
                provider=self._provider,
                model=model,
                input_text=input_text,
            )
        return trace_llm_call(
            lambda: raw,
            provider=self._provider,
            model=model,
            input_text=input_text,
        )
